runtime/interpreter/value_range.go

```diff
@@ -0,0 +1,226 @@
+			LocationRange: locationRange,
+			Message:       "step value cannot be zero",
+		})
+	}
+
+	// Validate that the sequence is moving towards the end value.
+	// If start < end, step must be > 0
+	// If start > end, step must be < 0
+	// If start == end, step doesn't matter.
+	if (start.Less(interpreter, end, locationRange) && step.Less(interpreter, zeroValue, locationRange)) ||
+		(start.Greater(interpreter, end, locationRange) && step.Greater(interpreter, zeroValue, locationRange)) {
+
+		panic(InclusiveRangeConstructionError{
+			LocationRange: locationRange,
+			Message: fmt.Sprintf(
+				"sequence is moving away from end: %s due to the value of step: %s and start: %s",
+				end,
+				step,
+				start,
+			),
+		})
+	}
+
+	return createInclusiveRange(interpreter, locationRange, start, end, step, rangeType, rangeSemaType)
+}
+
+func createInclusiveRange(
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	start IntegerValue,
+	end IntegerValue,
+	step IntegerValue,
+	rangeType InclusiveRangeStaticType,
+	rangeSemaType *sema.InclusiveRangeType,
+) *CompositeValue {
+	fields := []CompositeField{
+		{
+			Name:  sema.InclusiveRangeTypeStartFieldName,
+			Value: start,
+		},
+		{
+			Name:  sema.InclusiveRangeTypeEndFieldName,
+			Value: end,
+		},
+		{
+			Name:  sema.InclusiveRangeTypeStepFieldName,
+			Value: step,
+		},
+	}
+
+	rangeValue := NewCompositeValueWithStaticType(
+		interpreter,
+		locationRange,
+		nil,
+		rangeSemaType.QualifiedString(),
+		common.CompositeKindStructure,
+		fields,
+		common.ZeroAddress,
+		rangeType,
+	)
+
+	rangeValue.Functions = map[string]FunctionValue{
+		sema.InclusiveRangeTypeContainsFunctionName: NewHostFunctionValue(
+			interpreter,
+			sema.InclusiveRangeContainsFunctionType(
+				rangeSemaType.MemberType,
+			),
+			func(invocation Invocation) Value {
+				needleInteger := convertAndAssertIntegerValue(invocation.Arguments[0])
+
+				return rangeContains(
+					rangeValue,
+					rangeType,
+					invocation.Interpreter,
+					invocation.LocationRange,
+					needleInteger,
+				)
+			},
+		),
+	}
+
+	return rangeValue
+}
+
+func rangeContains(
+	rangeValue *CompositeValue,
+	rangeType InclusiveRangeStaticType,
+	interpreter *Interpreter,
+	locationRange LocationRange,
+	needleValue IntegerValue,
+) BoolValue {
+	start := getFieldAsIntegerValue(rangeValue, interpreter, locationRange, sema.InclusiveRangeTypeStartFieldName)
+	end := getFieldAsIntegerValue(rangeValue, interpreter, locationRange, sema.InclusiveRangeTypeEndFieldName)
+	step := getFieldAsIntegerValue(rangeValue, interpreter, locationRange, sema.InclusiveRangeTypeStepFieldName)
+
+	result := start.Equal(interpreter, locationRange, needleValue) ||
+		end.Equal(interpreter, locationRange, needleValue)
+
+	if result {
+		return TrueValue
+	}
+
+	greaterThanStart := needleValue.Greater(interpreter, start, locationRange)
+	greaterThanend := needleValue.Greater(interpreter, end, locationRange)
+
+	if greaterThanStart == greaterThanend {
```

Comment:
I was confused at first about what this condition is checking, and took me a while to realize. Maybe it would be more intuitive to have the two booleans checking `smallerThanStart` (or smallerOrEqualThanStart?) and `greaterThanEnd`, and have the condition as `if smallerThanStart || greaterThanEnd`.

Reply:
I don't think `if smallerThanStart || greaterThanEnd` will work since step can be negative also. So we need to check if needle is greater than one and smaller than another.

To improve the readability, I've extracted it to a separate function `isNeedleBetweenStartEndExclusive`. Hopefully it makes it a bit clearer.